### PR TITLE
Cleanup partially imported disks on CTRL+C

### LINKED ISSUE #651: disk requires special intervention to delete if upload process is Ctrl-C'd
### Component

- [X] CLI
- [ ] SDK
- [ ] Something else
- [ ] Not sure

### What was the expected behaviour

If the upload process is interrupted, the disk receiving the upload should be deleted.

### What is the current behaviour and what actions did you take to get there

I was uploading a disk with the CLI:

```
oxide disk import --path $HOME/oxide/iodriver.iso --disk iodriver-4096 --project iodriver --description "iodriver" --snapshot iodriver-r4-snapshot-4096 --image iodriver-4096 --image-description "iodriver" --image-os nixos --image-version 23.11 --disk-block-size 4096
```

I `ctrl-c`'d this part way through on accident, interrupting the upload. At this point, the disk in the web interface displayed as `IMPORTING_FROM_BULK_WRITES`, and I could not delete it.

To make the disk delible, I used `oxide api` to manually move it along:

```
# indicate writes have stopped
oxide api -X POST /v1/disks/9943337a-c7cd-4e7b-8945-60aa2d585c64/bulk-write-stop

# finalize my half-uploaded image to a snapshot
echo '{ "snapshot_name": "ugh" }' | oxide api -X POST /v1/disks/9943337a-c7cd-4e7b-8945-60aa2d585c64/finalize --input -
```

As shown here, you need to tell it writing is done, which puts it into the state of waiting for finalization. Then you need to finalize it, which results in a created **snapshot**. At this point, the disk moves back to the `detached` state, and you can go delete the snapshot this created as well. These are steps the import process would have done automatically after finishing the upload, if it had been allowed to finish.

### CLI or SDK version

Oxide CLI 0.4.0+20240327.0 Built from commit: a55649b70e749ed2cd51877240200f81b35d0d3c  Oxide API: 20240327.0

### Operating system

illumos

### Anything else you would like to add?

This is a known behavior, which has also been brought up in https://github.com/oxidecomputer/omicron/issues/2987, an issue which tracks the problem that there is no "cancel import" state transition for a disk.

In that thread, it's been suggested to, for now, handle Ctrl-C by finalizing the disk, and then deleting the resulting disk/snapshot, a sort of "graceful shutdown" in this case. It wouldn't handle harder interruptions (network drop, reboot, etc.), but ctrl-c is easiest instance of this to hit right now.

### PR DESCRIPTION
**Pull request by wfchandler**
If a `disk import` request is interrupted then the disk is likely to be
left in a `ImportReady` or `ImportingFromBulkWrites` state, which
prevents the user from deleting it. The commonly occurs when users
interrupt the request using ^C, and is a frustrating experience.

Add a new `GracefulShutdown` struct to listen for SIGINTs. This will
cancel the executing `Future` on interrupt and run a cleanup task which
will return an error, even if cleanup succeeds. A second SIGINT causes
the process to exit immediately.

Hook calls make in the critical section of the import into the new
system to ensure we delete the partially imported disk.

This introduces a race where the user may cancel the import operation
while the disk creation saga is still executing and the disk is in
`Creating` state. Attempting to finalize or delete the disk in this
state will fail, so we are forced to wait until it has transitioned to
`ImportReady` before proceeding. To avoid spamming the API too heavily,
we poll the disk state every 500ms for no more than ten tries.

The new polling creates a problem for our testing, though. We need the
initial disk API query to return a 404, but subsequent ones to find the
disk. `httpmock` does not provide a way to remove a mock after N hits,
and we have no graceful way of replacing the mock while running the
`oxide` binary. There is [an open issue](https://github.com/alexliesenfeld/httpmock/issues/96) to add an option like this to
`httpmock`, but for right now we have no option but to delete the tests
that check for failure.

Closes https://github.com/oxidecomputer/oxide.rs/issues/651

### CODE REVIEW

**Comment by wfchandler** [comment]
Recording:

https://github.com/user-attachments/assets/c6dccac1-95f6-49d6-914a-5af4bfac5d36

**Comment by wfchandler** [comment]
Opened https://github.com/alexliesenfeld/httpmock/pull/108 to add a `delete_after_n` method to `httpmock` that would allow us to clear the initial mock. In the meantime I've just removed the failing tests in this PR. We can add them back in if/when my change to `httpmock` is accepted.

File: cli/src/cmd_disk.rs [hunk 1]
```diff
@@ -435,6 +450,11 @@ impl crate::AuthenticatedCmd for CmdDiskImport {
 
             offset += CHUNK_SIZE;
             i += 1;
+
+            if shutdown.shutdown_requested()? {
+                self.remove_failed_disk(client).await?;
+                bail!("user canceled request");
```
Comment (david-crespo) [inline cli/src/cmd_disk.rs]: Would be nice to say something more concrete like "User canceled image upload"

**Comment by ahl** [comment]
I had been moving the disk upload functionality to the SDK. This seems to complicate that. I had imagined we would do cleanup from a custom `Drop` implementation (albeit synchronously in the absence of `async` `Drop`). When we receive a SIGINT we could cancel the future and let the Drop logic do the cleanup.

From a scan through your changes, it seems like there may be obstacles in the API such as needing to wait for particular state transitions. There may be situations where we want to change the API or state machine.

File: cli/src/shutdown.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        let (tx, rx) = tokio::sync::watch::channel(());
+
+        tokio::spawn(async move {
+            let mut force_shutdown = false;
+            loop {
+                signal::ctrl_c().await.expect("Failed to listen for CTRL+C");
+
+                if force_shutdown {
+                    eprintln_nopipe!("Shutting down immediately.\n{exit_msg}.");
+                    std::process::exit(130);
```
Comment (jmpesp) [inline cli/src/shutdown.rs:29]: Where did this value come from?

**Review by jmpesp — APPROVED** [review]
These changes look good to me, though Adam's review should supersede this one. One comment from me:

File: cli/src/shutdown.rs (hunk 2 above)
Comment (wfchandler) [reply cli/src/shutdown.rs:29]: This is matching the [exit code `bash` would use](https://github.com/bminor/bash/blob/142bbdd89e4d5bb62aea4469d1d2c24cf470afd8/jobs.c#L2757) had the standard `SIGINT` handler been invoked. It's not exactly the same behavior as `WIFSIGNALED` won't be aware of the signal, but I think it's close enough for our purposes.

**Comment by wfchandler** [comment]
> I had been moving the disk upload functionality to the SDK. This seems to complicate that. I had imagined we would do cleanup from a custom `Drop` implementation (albeit synchronously in the absence of `async` `Drop`). When we receive a SIGINT we could cancel the future and let the Drop logic do the cleanup.
> 
> From a scan through your changes, it seems like there may be obstacles in the API such as needing to wait for particular state transitions. There may be situations where we want to change the API or state machine.

@ahl I'm not sure where that leaves this PR. Should I close it, or do we expect the work to update the API to take long enough it's worthwhile merging this as a stopgap.

**Comment by ahl** [comment]
> > I had been moving the disk upload functionality to the SDK. This seems to complicate that. I had imagined we would do cleanup from a custom `Drop` implementation (albeit synchronously in the absence of `async` `Drop`). When we receive a SIGINT we could cancel the future and let the Drop logic do the cleanup.
> > From a scan through your changes, it seems like there may be obstacles in the API such as needing to wait for particular state transitions. There may be situations where we want to change the API or state machine.
> 
> @ahl I'm not sure where that leaves this PR. Should I close it, or do we expect the work to update the API to take long enough it's worthwhile merging this as a stopgap.

@wfchandler I was hoping we might change the approach here to do cleanup when we drop the future. Do you think that's viable?

**Comment by wfchandler** [comment]
> > > I had been moving the disk upload functionality to the SDK. This seems to complicate that. I had imagined we would do cleanup from a custom `Drop` implementation (albeit synchronously in the absence of `async` `Drop`). When we receive a SIGINT we could cancel the future and let the Drop logic do the cleanup.
> > > From a scan through your changes, it seems like there may be obstacles in the API such as needing to wait for particular state transitions. There may be situations where we want to change the API or state machine.
> > 
> > 
> > @ahl I'm not sure where that leaves this PR. Should I close it, or do we expect the work to update the API to take long enough it's worthwhile merging this as a stopgap.
> 
> @wfchandler I was hoping we might change the approach here to do cleanup when we drop the future. Do you think that's viable?

@ahl I think the latest revision of this PR should be compatible with your SDK changes. We can replace [self.upload_disk](https://github.com/oxidecomputer/oxide.rs/blob/c63efaeea9bdf163b6b8cc10fbc67ad539164c09/cli/src/cmd_disk.rs#L541-L549) with the equivalent SDK command when that's available. WDYT?

**Comment by david-crespo** [comment]
Ran into this today. Would be nice to figure out how to get this in. In the meantime I think I’m going to add stop and and finalize to the web console.

**Comment by wfchandler** [comment]
> Ran into this today. Would be nice to figure out how to get this in. In the meantime I think I’m going to add stop and and finalize to the web console.

I have almost finished a rewrite of this that adds a convenience wrapper to the SDK, hopefully ready to go later today.

**Comment by wfchandler** [comment]
Closing in favor of https://github.com/oxidecomputer/oxide.rs/pull/840

**Pull request closed without merging**